python_scripts/clustering_kmeans.py

```diff
@@ -424,59 +406,21 @@ def plot_silhouette_scores(
 # %% [markdown]
 #
 # This plot seems to be very similar to the pairplot we obtained with the 6
-# clusters found by k-means on our preprocessed data. Note that the colors are
-# different, because the ordering of the labels is arbitrary (both for the
-# k-means cluster and the manually assigned labels). But the way of grouping
-# the data points look similar.
-#
-# Let's quantify the agreement between the clusters found by k-means and the
-# combination of the "Species" and "Sex" labels using the [Normalized Mutual
-# Information](https://scikit-learn.org/stable/modules/clustering.html#mutual-info-score)
-# (NMI) score.
-
-# %%
-from sklearn.metrics.cluster import normalized_mutual_info_score
-
-nmi = normalized_mutual_info_score(
-    species_and_sex_labels,
-    cluster_labels,
-)
-print(f"Normalized Mutual Information (NMI): {nmi:.3f}")
-
-# %% [markdown]
-#
-# This value is very close to 1.0, which indicates a very strong agreement.
+# clusters found by k-means on our preprocessed data, i.e. in both cases plots
+# that display 3 clusters can be further divided into a group of proportionally
+# smaller penguins. Only the colors may differ, as the ordering of the labels is
+# arbitrary (both for the k-means cluster and the manually assigned labels).
 #
 # The conclusion is that we relate the clusters found by running k-means on
 # those preprocessed features to a meaningful (human) way to partition the
-# penguins records.
-#
-# Note however that this is not always the case. For **k-means to yield
-# meaningful results, the data must be have an approximately balanced, convex
-# and isotropic cluster structure** after preprocessing. That is, the clusters
-# must have a spherical shape in the feature space and approximately the same
-# size.
+# penguins records. Notice however that this may not always be the case.
 #
 # We cannot stress enough that the choice of the features and preprocessing
 # steps are crucial: if we had not standardized the numerical data, or we had
-# not included the "Sex" feature or if we had scaled its one-hot encoding by a
+# not included the "Sex" feature, or if we had scaled its one-hot encoding by a
 # factor of 10, we would probably not have been able to discover interpretable
 # clusters.
 #
 # Furthermore, **many natural datasets would not satisfy the k-means
-# assumptions** even after non-trivial preprocessing. In those cases, we can
-# either try alternatives to k-means that favor different cluster shapes (for
-# instance HDBSCAN or Gaussian Mixture Models) or we can try to isolate
-# row-wise or column-wise subsets of the data that are more likely to exhibit a
-# cluster structure. Or sometimes, we can decide to partition the data with
-# k-means with a large number of clusters, even if they are not interpretable
-# and use the distance to centroids as preprocessing for another task. It all
-# depends on the specific application domain and the downstream use of the
-# resulting clusters.
```

Comment:
If we do not include this paragraph here, I think we should instead have a slide that conveys this message.